### ISSUE TITLE
Allow to select multiple waypoints from different rows

### ISSUE THREAD

**Issue by morevnaproject** [synfig/synfig#762]
Allow user to select multiple waypoints by holding CTRL.

Currently this is possible if waypoints belong to the same parameter.
But if waypoints belong to different parameters, then it is impossible to select them together.

**Comment by yshkp** [comment]
@morevnaproject As it is related to my project which I proposed to you so you can assign it to me and further it will be continued during our project period.

**Comment by morevnaproject** [comment]
@yshkp Assigned to you now. ^__^

**Comment by yshkp** [comment]
@morevnaproject I will be submitting you a draft proposal tomorrow which will also mention this issue about the approach I thought. If you allow me I can also include this is in my calendar of coding period. 
I have also thought some improvements regarding waypoints handling which definitely improve it.

**Comment by morevnaproject** [comment]
@yshkp Thanks for keeping me updated! Looking forward for your draft. ^__^
P.S. Please send it to me via email.

**Comment by yshkp** [comment]
@morevnaproject Please check your mail. I sent you the link to my draft proposal. Please take a look and review. Any suggestions would be appreciated.
I hope it matches to your expectations. ^__^

**Comment by morevnaproject** [comment]
Got your mail, thank you! ^__^

**Comment by yshkp** [comment]
@morevnaproject As I have started working on the task that you gave me and on the early basis I came to an idea that: If it is allowed to select multiple rows which are rendered using cell renderer then automatically user can select multiple waypoints using CTRL key as the main issue was due to operating on a single row.
Please tell me if I am on the right track.

**Comment by morevnaproject** [comment]
It's hard to tell without trying. I doubt this will work, but still you can try to enable multiple selection.

Again, the safest way is to re-write the widget completely. For reference on writing custom widgets you can check TimeSlider widget (gray area on top of TimeTrack Panel) - https://github.com/synfig/synfig/blob/master/synfig-studio/src/gui/widgets/widget_timeslider.cpp

Graps Panel is also a good example of custom widget related with time/parameters.

**Comment by yshkp** [comment]
@morevnaproject Well, yes now I am going to implement it in re-writing the widget.
Just gone through some docs so I thought this could work so suggested you an idea. But it's good to have a custom widget so without wasting time now working on it. :)

**Comment by morevnaproject** [comment]
Also, I suggest to create a copy of TImeTrack panel and do all work there. You can call it "TimeTrack-2". This will allow incremental development, without breaking existing functionality. ^__^

**Comment by yshkp** [comment]
@morevnaproject Do we need to continue with TreeView again in time track panel ?

**Comment by morevnaproject** [comment]
If there will be a new widget, than there will be no need for TreeView there.

**Comment by yshkp** [comment]
@morevnaproject I have gone through all the containers to find best suitable for time track panel and as per the references you gave there also DrawingArea is used which gives freedom to use it in our desired way. So I think it can be used in time track too.
What's your opinion on this?

**Comment by morevnaproject** [comment]
@yshkp Yes, DrawingArea is a right choice for this task. ^__^

**Comment by yshkp** [comment]
@morevnaproject I checked `CellRenderer_TimeTrack` that it is inherited from `Gtk::CellRenderer` and also a custom renderer to render the param model in TimeTrack. So I think CellRenderer would work for DrawingArea also. I created a context in TimeTrack to get in that environment. For TreeView Columns, I think it would be done separately.

**Comment by yshkp** [comment]
@morevnaproject We can replace TreeView columns with the rectangle widget which itself be in the context in this way we can achieve the desired operation.

**Comment by morevnaproject** [comment]
Can you please clarify which rectangle widget you referring to? ^_^

**Comment by yshkp** [comment]
@morevnaproject I am referring to `Gdk::Rectangle`, so that focus can be given and also allowed to edit.

**Comment by morevnaproject** [comment]
Now, question TreeView Columns - are you referring to this? - https://github.com/synfig/synfig/blob/master/synfig-studio/src/gui/docks/dock_timetrack.cpp#L103

**Comment by yshkp** [comment]
@morevnaproject Yes

**Comment by morevnaproject** [comment]
You do not to replace it with anything then. Currently TimetrackView is implemented as TreeView with a single column (there is actually a second column, but it is empty, used  height alignment). You just need to replace whole TreeView with DrawingArea.

**Comment by yshkp** [comment]
@morevnaproject Yup I am on it and in process of replacing TreeView with DrawingArea,objects of TreeView is used so for that also appropriate replacement is needed in accordance with DrawingArea.

**Comment by morevnaproject** [comment]
What I want to say is that you do not need to replace TreeViewColumn with anything. TreeViewColumn is a part of TreeView's internal structure. ^__^

**Comment by yshkp** [comment]
Oh yes I got it before just saying for other objects. 😅

**Comment by yshkp** [comment]
@morevnaproject I am almost there to replace it with DrawingArea but few issues like in replacing TreeView with DrawingArea, some TreeView methods go out of scope like for appending TreeView column, converting the coordinates etc. so for that, are custom methods also need to build?

**Comment by morevnaproject** [comment]
Consider that Drawing Area doesn't needs a method for appending TreeView column, since it doesn't needs column at all. For converting coordinates please check TimeSlider widget (gray area on top of TimeTrack Panel) - https://github.com/synfig/synfig/blob/master/synfig-studio/src/gui/widgets/widget_timeslider.cpp

**Comment by yshkp** [comment]
Oh, the thing not to replace TreeView column made me to think for its append. As earlier too, I was on the process for the removal of column.

**Comment by yshkp** [comment]
@morevnaproject But one thing that if TreeView column is removed then `CellRender_TimeTrack` would not work also, so another way is required to render the value nodes.

**Comment by morevnaproject** [comment]
Exactly. You have to write a code which will display waypoints in DrawingArea.

**Comment by yshkp** [comment]
@morevnaproject Then, Do I need to modify `CellRenderer_TimeTrack` also?  
As it is responsible for displaying waypoints.

**Comment by morevnaproject** [comment]
I think your problem is that you are trying to "map" features of current implementation (TreeView) to new implementation (DrawingArea) and trying to make it work "in one shot".

Instead I suggest to go with incremental changes. By "rewriting TimeTrack" widget I mean you should take DrawingArea widget and try to implement all UI functionality "from scratch".
And this can be done in incremental steps.

I.e.:

1.  Create a panel with empty drawing area.

2. Write a code to make Drawing Area to display rows, corresponding to each parameter.
![screenshot_004](https://user-images.githubusercontent.com/332868/55680867-a0fe0f00-5949-11e9-9143-dad79223a256.png)

3. After that, make Drawing Area to display vertical lines, corresponding to frames. You will not need that in final widget, but this will help you to see layout of waypoints during development.

![screenshot_004](https://user-images.githubusercontent.com/332868/55680916-529d4000-594a-11e9-8c62-5c06daa47eb4.png)

This code can be borrowed from TimeSlider widget (gray/green area on top of TimeTrack Panel), which is also based on Drawing Area.

![screenshot_005](https://user-images.githubusercontent.com/332868/55680940-a90a7e80-594a-11e9-9643-3f6976098bc9.png)

4. Make DrawingArea react on timeline zoom - when you hover on TimeSlider widget and scroll mouse wheel it changes horizontal scale for both TimeSlider and TimeTrack. This should be implemented accordingly - vertical lines you made on step 3 should react on that.

This code also can be borrowed from TimeSlider widget.

5. Write a code which displays in Drawing Area red circles for waypoints.

![screenshot_004](https://user-images.githubusercontent.com/332868/55680971-2209d600-594b-11e9-8d1f-78c58bfc0dad.png)

6. Replace red circles with real waypoint icons.

7. Make waypoints clickable (selectable)

8. Allow to move selected waypoints (this is where you can borrow code from old TimeSlider based on TreeView).

9. Write a code to allow select multiple waypoints.

10. Modify code to process moving of several waypoints.

**Comment by yshkp** [comment]
@morevnaproject  Thanks a lot for this reference. ^__^
To maintain its functionality as before I was trying to make all methods similar but in a new way so that it can be handled in DrawingArea. 
But now no more queries as this reference you gave a clear view of what to do.

**Comment by yshkp** [comment]
@morevnaproject To get the info of parameters of a certain layer, Can `LayerParamModel` help in this`?
Due to the institute's work, I am not able to give proper time but trying to do as fast I can.

**Comment by yshkp** [comment]
@morevnaproject Up to step 4 is completed, now to render waypoints I think I should take this from the method in `CellRenderer_TimeTrack` and then display them as red circles. Please let me know your opinion.
P.S. - I found that keyframe list block size is not proper. What could be the reason for this?

**Comment by morevnaproject** [comment]
> Up to step 4 is completed, now to render waypoints I think I should take this from the method in CellRenderer_TimeTrack and then display them as red circles.

Yes, I advise to look at ```CellRenderer_TimeTrack::render_vfunc```.

> P.S. - I found that keyframe list block size is not proper. What could be the reason for this?

Sorry, I do not understand what you mean. Can you please describe in more detail? ^__^

**Comment by yshkp** [comment]
> Sorry, I do not understand what you mean. Can you please describe in more detail? ^__^
No worries, I found the solution and fixed it. ^__^

**Comment by yshkp** [comment]
@morevnaproject The rows which are needed to draw corresponding to parameters, I want to implement it as by counting total parameters and then loop it in timetrack to draw horizontal lines. Is it good to have like this?

**Comment by morevnaproject** [comment]
Yes, this is a good way to go! ^__^

**Comment by yshkp** [comment]
@morevnaproject I tried to implement as per the way suggested but an error of segmentation fault is occurring. 
I created the iterator to iterate through the children of a `LayerParamTreeStore` model and thus counting its iteration.

**Comment by yshkp** [comment]
@morevnaproject Please check mail when you get time. ^__^

**Comment by morevnaproject** [comment]
Hi @yshkp  ! Do you plan to continue working on this?
